openage/convert/smp.pyx

```diff
@@ -0,0 +1,837 @@
+        return
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef numpy.ndarray determine_rgba_matrix(vector[vector[pixel]] &image_matrix,
+                                         main_palette, player_palette):
+    """
+    converts a palette index image matrix to an rgba matrix.
+    """
+
+    cdef size_t height = image_matrix.size()
+    cdef size_t width = image_matrix[0].size()
+
+    cdef numpy.ndarray[numpy.uint8_t, ndim=3] array_data = \
+        numpy.zeros((height, width, 4), dtype=numpy.uint8)
+
+    # micro optimization to avoid call to ColorTable.__getitem__()
+    cdef list m_lookup = main_palette.palette
+    cdef list p_lookup = player_palette.palette
+
+    cdef uint8_t r
+    cdef uint8_t g
+    cdef uint8_t b
+    cdef uint8_t a
+
+    cdef vector[pixel] current_row
+    cdef pixel px
+    cdef pixel_type px_type
+    cdef int px_index
+    cdef int px_palette
+
+    cdef size_t x
+    cdef size_t y
+
+    for y in range(height):
+
+        current_row = image_matrix[y]
+
+        for x in range(width):
+            px = current_row[x]
+            px_type = px.type
+            px_index = px.index
+            px_palette = px.palette
+
+            if px_type == color_standard:
+                # look up the palette secition
+                # palettes have 1024 entries
+                # divided into 4 sections
+                palette_section = px_palette % 4
+
+                # the index has to be adjusted
+                # to the palette section
+                index = px_index + 256 * palette_section
+
+                # look up the color index in the
+                # main graphics table
+                r, g, b, alpha = m_lookup[index]
+
+                # TODO: alpha values are unused
+                # in 0x0C and 0x0B version of SMPs
+                alpha = 255
+
+            elif px_type == color_transparent:
+                r, g, b, alpha = 0, 0, 0, 0
+
+            elif px_type == color_shadow:
+                r, g, b, alpha = 0, 0, 0, px_index
+
+            else:
+                if px_type == color_player:
+                    alpha = 255
+
+                elif px_type == color_outline:
+                    alpha = 254
+
+                else:
+                    raise ValueError("unknown pixel type: %d" % px_type)
+
+                # get rgb base color from the color table
+                # store it the preview player color
+                # in the table: [16*player, 16*player+7]
+                r, g, b = p_lookup[px_index]
+
+            # array_data[y, x] = (r, g, b, alpha)
+            array_data[y, x, 0] = r
+            array_data[y, x, 1] = g
+            array_data[y, x, 2] = b
+            array_data[y, x, 3] = alpha
+
+    return array_data
+
+cdef (uint8_t,uint8_t) get_palette_info(pixel image_pixel):
+    """
+    returns the palette used for a pixel.
+    """
+    return image_pixel.palette >> 4, image_pixel.palette & 0x03
```

Comment:
two numbers are returned, i'd welcome a comment what both of the numbers mean and why there are two.

Reply:
Note to myself: the function should return `image_pixel.palette >> 2`